easy_transformer/hook_points.py

```diff
@@ -1,149 +1,216 @@
+        if not isinstance(direction, HookDirection):
+            raise ValueError(f"Invalid direction {direction}")
+        if (direction == HookDirection.FORWARD) or (direction == HookDirection.BOTH):
+            for hook in self.forward_hooks:
                 hook.remove()
-            self.fwd_hooks = []
-        if (dir == "bwd") or (dir == "both"):
-            for hook in self.bwd_hooks:
+            self.forward_hooks = []
+        if (direction == HookDirection.BACKWARD) or (direction == HookDirection.BOTH):
+            for hook in self.backward_hooks:
                 hook.remove()
-            self.bwd_hooks = []
-        if dir not in ["fwd", "bwd", "both"]:
-            raise ValueError(f"Invalid direction {dir}")
+            self.backward_hooks = []
 
     def clear_context(self):
-        del self.ctx
-        self.ctx = {}
+        del self.context
+        self.context = {}
 
     def forward(self, x):
+        """Identity function"""
         return x
 
     def layer(self):
-        # Returns the layer index if the name has the form 'blocks.{layer}.{...}'
-        # Helper function that's mainly useful on EasyTransformer
-        # If it doesn't have this form, raises an error -
-        split_name = self.name.split(".")
-        return int(split_name[1])
+        """
+        Returns the layer index if the name has the form 'blocks.{layer}.{...}'
+        Helper function that's mainly useful on EasyTransformer
+        If it doesn't have this form, raises an error -
+        """
+        _, layer, _ = self.name.split(".")
+        return int(layer)
 
 
 # %%
+# TODO maybe this should be a decorator so clients don't have to call setup()?
+# or just a class should be enough?
 class HookedRootModule(nn.Module):
     """
     A class building on nn.Module to interface nicely with HookPoints
     Adds various nice utilities, most notably run_with_hooks to run the model with temporary hooks, and run_with_cache to run the model on some input and return a cache of all activations
 
+    See [EasyTransformer_Demo.ipynb] for an example of how to use it.
+
     WARNING: The main footgun with PyTorch hooking is that hooks are GLOBAL state. If you add a hook to the module, and then run it a bunch of times, the hooks persist. If you debug a broken hook and add the fixed version, the broken one is still there. To solve this, run_with_hooks will remove hooks at the start and end by default, and I recommend using reset_hooks liberally in your code.
 
     The main time this goes wrong is when you want to use backward hooks (to cache or intervene on gradients). In this case, you need to keep the hooks around as global state until you've run loss.backward() (and so need to disable the reset_hooks_end flag on run_with_hooks)
     """
 
-    def __init__(self, *args):
+    def __init__(self):
         super().__init__()
-        self.is_caching = False
+        self.is_caching = False  # TODO what's this used for?
```

Comment:
Deprecated, can delete

Reply:
wait, what about [this helpful warning? ](https://github.com/neelnanda-io/Easy-Transformer/blob/a540929a0ca34e321528f5e8b7716e30a0744ca9/easy_transformer/hook_points.py#L223)